BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java

```diff
@@ -7,88 +7,129 @@
 	}
 
 	public void delete(BlueKey key) throws BlueDbException {
-		TreeMap<BlueKey, BlueEntity<T>> data = load();
+		TreeMap<BlueKey, BlueEntity<T>> data = load(key.getGroupingNumber());
 		data.remove(key);
+		String pathString = Paths.get(path.toString(), String.valueOf(key.getGroupingNumber())).toString();
 		Blutils.save(pathString, data, serializer);
 	}
 
 	public BlueEntity<T> read(BlueKey key) throws BlueDbException {
-		Collection<BlueEntity<T>> values = load().values();
+		Collection<BlueEntity<T>> values = load(key.getGroupingNumber()).values();
 		for (BlueEntity<T> entity: values) {
 			if (entity.getKey().equals(key))
 				return entity;
 		}
 		return null;
 	}
 
-	public List<BlueEntity<T>> read() throws BlueDbException {
-		List<BlueEntity<T>> results = new ArrayList<>(load().values());
-		return results;
-	}
-
 	public List<BlueEntity<T>> read(long minTime, long maxTime) throws BlueDbException {
 		List<BlueEntity<T>> results = new ArrayList<>();
-		for (BlueEntity<T>entity: load().values()) {
+		List<Long> groupingNumbers = getAllFiles().stream()
+				.filter((f) -> isFileNameALong(f))
+				.map((f) -> f.getName())
+				.map((filename) -> Long.valueOf(filename))
+				.collect(Collectors.toList());
+		for (long groupingNumber: groupingNumbers) {
+			TreeMap<BlueKey, BlueEntity<T>> data = load(groupingNumber);
+			for (BlueEntity<T>entity: data.values()) {
 			if (Blutils.meetsTimeConstraint(entity.getKey(), minTime, maxTime))
 				results.add(entity);
+			}
 		}
 		return results;
 	}
 
+	
+	// TODO refactor and test and maybe move somewhere else
+	protected static boolean isFileNameALong(File file) {
+		try {
+			String fileName = file.getName();
+			long fileNameAsLong = Long.valueOf(fileName);
+			return true;
+		} catch(Exception e) {
+			return false;
+		}
+	}
+
+	// TODO refactor and test and maybe move somewhere else
+	protected static boolean isFileNameALongInRange(File file, long minValue, long maxValue) {
+		try {
+			String fileName = file.getName();
+			long fileNameAsLong = Long.valueOf(fileName);
+			return fileNameAsLong >= minValue && fileNameAsLong <= maxValue;
+		} catch(Exception e) {
+			return false;
+		}
+	}
+
+	// TODO refactor and test
+	private List<File> getAllFiles() {
+		File segmentFolder = path.toFile();
+		if (!segmentFolder.exists()) {
+			return new ArrayList<>();
+		} else {
+			return Arrays.asList(segmentFolder.listFiles());
+		}
+	}
+
 	@SuppressWarnings("unchecked")
-	private TreeMap<BlueKey, BlueEntity<T>> load() throws BlueDbException {
+	private TreeMap<BlueKey, BlueEntity<T>> load(long groupingNumber) throws BlueDbException {
+		Path groupPath = Paths.get(path.toString(), String.valueOf(groupingNumber));
 		try {
-			File file = new File(pathString);
+			File file = groupPath.toFile();
 			if (!file.exists()) {
 				return new TreeMap<>();
 			} else {
-				byte[] bytes = Files.readAllBytes(Paths.get(pathString));
+				byte[] bytes = Files.readAllBytes(groupPath);

```

Comment:
We could probably use another method in serializer called "readObjectFromFile" that everything can easily use

Reply:
I'd like to punt this to FileManager, which is probably what I'll implement next.